core/src/main/java/org/everit/email/store/ri/EmailStoreImpl.java

```diff
@@ -0,0 +1,688 @@
+import org.everit.blobstore.Blobstore;
+import org.everit.email.Attachment;
+import org.everit.email.Email;
+import org.everit.email.EmailAddress;
+import org.everit.email.HtmlContent;
+import org.everit.email.Recipients;
+import org.everit.email.store.EmailStore;
+import org.everit.email.store.ri.schema.qdsl.QAddress;
+import org.everit.email.store.ri.schema.qdsl.QAttachment;
+import org.everit.email.store.ri.schema.qdsl.QBinaryContent;
+import org.everit.email.store.ri.schema.qdsl.QEmail;
+import org.everit.email.store.ri.schema.qdsl.QEmailRecipient;
+import org.everit.email.store.ri.schema.qdsl.QHtmlContent;
+import org.everit.email.store.ri.schema.qdsl.QInlineImage;
+import org.everit.email.store.ri.schema.qdsl.QTextContent;
+import org.everit.persistence.querydsl.support.QuerydslSupport;
+import org.everit.transaction.propagator.TransactionPropagator;
+
+import com.mysema.query.Tuple;
+import com.mysema.query.sql.SQLQuery;
+import com.mysema.query.sql.dml.SQLDeleteClause;
+import com.mysema.query.sql.dml.SQLInsertClause;
+import com.mysema.query.types.Projections;
+
+/**
+ * Reference implementation of {@link EmailStore}.
+ */
+public class EmailStoreImpl implements EmailStore {
+
+  /**
+   * Types of recipient.
+   */
+  private enum RecipientType {
+    BCC, CC, FROM, TO
+  }
+
+  private static final int BUFFER_SIZE = 1024;
+
+  private static final int START_INDEX = 0;
+
+  private final Blobstore blobstore;
+
+  private final QuerydslSupport querydslSupport;
+
+  private final TransactionPropagator transactionPropagator;
+
+  /**
+   * Simple constructor.
+   *
+   * @param querydslSupport
+   *          a {@link QuerydslSupport} instance.
+   * @param transactionPropagator
+   *          a {@link TransactionPropagator} instance.
+   * @param blobstore
+   *          a {@link Blobstore} instance.
+   */
+  public EmailStoreImpl(final QuerydslSupport querydslSupport,
+      final TransactionPropagator transactionPropagator, final Blobstore blobstore) {
+    Objects.requireNonNull(querydslSupport, "The querydslSupport cannot be null!");
+    Objects.requireNonNull(transactionPropagator, "The transactionPropagator cannot be null!");
+    Objects.requireNonNull(blobstore, "The blobstore cannot be null!");
+
+    this.querydslSupport = querydslSupport;
+    this.transactionPropagator = transactionPropagator;
+    this.blobstore = blobstore;
+  }
+
+  private Attachment createAttachment(final Long blobId, final String contentType,
+      final String name) {
+    BlobInputStreamSupplier inputStreamSupplier = null;
+    if (blobId != null) {
+      inputStreamSupplier = new BlobInputStreamSupplier(blobstore, blobId);
+    }
+    Attachment attachment = new Attachment()
+        .withContentType(contentType)
+        .withName(name)
+        .withInputStreamSupplier(inputStreamSupplier);
+    return attachment;
+  }
+
+  private Long createBlob(final byte[] contentBytes) {
+    if ((contentBytes == null) || (contentBytes.length < 1)) {
```

Comment:
contentBytes cannot be null. Why do we have two functions to create blob? Why not just call forward to the one that uses InputStream by passing a ByteArrayInputStream?